kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/collections/CollectionMatchers.kt

```diff
@@ -176,41 +175,77 @@ fun <T> matchEach(expected: List<T>, asserter: (T, T) -> Unit): Matcher<Collecti
       }
    })
 
+private sealed interface MatchEachProblem {
+   val atIndex: Int
+   val shortMessage: String?
+   val longMessage: String?
+
+   data class SizeMismatch(override val atIndex: Int, val problem: String) : MatchEachProblem {
+      override val shortMessage: String
+         get() = problem
+      override val longMessage: String
+         get() = problem
+   }
+
+   data class Thrown(override val atIndex: Int, val problem: Throwable) : MatchEachProblem {
+      override val shortMessage: String?
+         get() = problem.message
+      override val longMessage: String?
+         get() = problem.stackTraceToString()
+   }
+}
+
 /**
  * Asserts that each element in the collection matches its corresponding matcher in [assertions].
  * Elements will be compared sequentially in the order given by the iterators of the collections.
  */
 fun <T> matchEach(assertions: List<(T) -> Unit>): Matcher<Collection<T>?> = neverNullMatcher { actual ->
-   data class MatchEachProblem(val atIndex: Int, val problem: String?)
-
    val problems = errorCollector.runWithMode(ErrorCollectionMode.Hard) {
       actual.mapIndexedNotNull { index, element ->
          if (index !in assertions.indices) {
-            MatchEachProblem(
+            MatchEachProblem.SizeMismatch(
                index,
                "Element has no corresponding assertion. Only ${assertions.size} assertions provided"
             )
          } else {
             runCatching {
                assertions[index](element)
             }.exceptionOrNull()?.let { exception ->
-               MatchEachProblem(index, exception.message)
+               MatchEachProblem.Thrown(index, exception)
             }
          }
       }
    } + (actual.size until assertions.size).map {
-      MatchEachProblem(
+      MatchEachProblem.SizeMismatch(
          it,
          "No actual element for assertion at index $it"
       )
    }
 
-   MatcherResult(
-      problems.isEmpty(),
-      {
-         "Expected each element to pass its assertion, but found issues at indexes: [${problems.joinToString { it.atIndex.toString() }}]\n\n" +
-            problems.joinToString(separator = "\n") { "${it.atIndex} => ${it.problem}" }
-      },
-      { "Expected some element to fail its assertion, but all passed." },
-   )
+   when (problems.size) {
+      0 -> MatcherResult(
+         true,
+         { "" },
+         { "Expected some element to fail its assertion, but all passed." },
+      )
+
+      1 -> {
+         val problem = problems.first()
+         MatcherResultWithError(
+            (problem as? MatchEachProblem.Thrown)?.problem,
```

Comment:
Can we be sure it's not a single SizeMismatch?

e.g. 
```kotlin
emptyList<Int>().shouldMatchEach({ it shouldBe 5})
```

Would probably result in problems being a single SizeMismatch element?

Reply:
If so, then the expression `(problem as? MatchEachProblem.Thrown)?.problem` is evaluated as `null`, which is ok for `MatcherResultWithError` I suppose.